### PR TITLE
Add annotation processor for content resolver

### LINKED ISSUE #443: Annotation processor for generating StorIOContentResolver type mapping

### PR DESCRIPTION
**Pull request by nikitin-da**
Closes #443 
@artem-zinnatullin PTAL, please =)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -71,12 +71,14 @@ ext.libraries = [
...
-        storIOContentResolver          : project(':storio-content-resolver'),
-        storIOSQLiteAnnotation         : project(':storio-sqlite-annotations'),
-        storIOSQLiteAnnotationProcessor: project(':storio-sqlite-annotations-processor'),
-        storIOTestCommon               : project(':storio-test-common'),
+        storIOCommon                             : project(':storio-common'),
+        storIOSQLite                             : project(':storio-sqlite'),
+        storIOContentResolver                    : project(':storio-content-resolver'),
+        storIOSQLiteAnnotation                   : project(':storio-sqlite-annotations'),
+        storIOSQLiteAnnotationProcessor          : project(':storio-sqlite-annotations-processor'),
+        storIOContentResolverAnnotation          : project(':storio-content-resolver-annotations'),
```
Comment (artem-zinnatullin) [inline build.gradle]: Can you please rename it to `storIOContentResolverAnnotations` "s" at the end, my bad that I forgot to do it for sqlite annotations

File: build.gradle [hunk 2]
```diff
@@ -71,12 +71,14 @@ ext.libraries = [
...
+        storIOContentResolverAnnotationProcessor : project(':storio-content-resolver-annotations-processor'),
```
Comment (artem-zinnatullin) [inline build.gradle]: same here

File: storio-content-resolver-annotations-processor/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/processor/StorIOContentResolverProcessor.java [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    public SourceVersion getSupportedSourceVersion() {
+        return SourceVersion.latestSupported();
+    }
+
+    //endregion
+
+    /**
+     * For those who don't familiar with Annotation Processing API — this is the main method of Annotation Processor lifecycle
+     * <p>
+     * It will be after Java Compiler will find lang elements annotated with annotations from {@link #getSupportedAnnotationTypes()}
```
Comment (artem-zinnatullin) [inline storio-content-resolver-annotations-processor/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/processor/StorIOContentResolverProcessor.java]: It will be `called` after Java Compiler (looks like I made same mistake in javadoc of sqlite-ap) :)

**Comment by artem-zinnatullin** [comment]
@nikitin-da sorry, way too busy this week… will finish review soon!

**Comment by artem-zinnatullin** [comment]
Looks like we can share a lot of code between annotation processors, I'll think about it.

**Comment by nikitin-da** [comment]
Nice idea! I'll move it to :storio-common-annotations-processor

**Comment by artem-zinnatullin** [comment]
> I'll move it to :storio-common-annotations-processor

:+1:

File: storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+@Target(TYPE)
+@Retention(RUNTIME) // we allow users to write reflection based code to work with annotation
+public @interface StorIOContentResolverType {
+
+    /**
+     * Required: Specifies uri
+     *
+     * @return uri
+     */
+    String uri();
```
Comment (artem-zinnatullin) [inline storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java]: `@NonNull` (you can add dependency on support-annotations)

File: storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverColumn.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+@Target(FIELD)
+@Retention(RUNTIME)
+public @interface StorIOContentResolverColumn {
+
+    /**
+     * Required: specifies column name
+     *
+     * @return non-null column name
+     */
+    String name();
```
Comment (artem-zinnatullin) [inline storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverColumn.java]: `@NonNull`

File: storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java (hunk 4 above)
Comment (nikitin-da) [inline storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java]: Ook

File: storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * Annotation for marking class as StorIOContentResolverType to generate object mapping code
+ */
+@Target(TYPE)
+@Retention(RUNTIME) // we allow users to write reflection based code to work with annotation
+public @interface StorIOContentResolverType {
```
Comment (artem-zinnatullin) [inline storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java]: I think we should also add ability to set different `uris` for `query`, `insert`, `update`, `delete`, I know several projects that use different uris for different operations :crying_cat_face:

Comment (artem-zinnatullin) [inline storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java]: Let's do it in a separate pr!

**Comment by artem-zinnatullin** [comment]
:+1:, nullability and rebase on latest master and I'll merge it!

Then will extract common code between annotations processors in a separate PR!

File: storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java (hunk 6 above)
Comment (nikitin-da) [inline storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java]: Ook, but should we provide the default `uri` for all operations? if user doesn't want to set it separately?
Also, it can allow to make 1.7 without braking backward campatibility

Comment (artem-zinnatullin) [inline storio-content-resolver-annotations/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/StorIOContentResolverType.java]: I guess yes, only `uri` should be required by default and user may override `uri` for some operations and annotation processor should use overriden variant if it set.

**Comment by nikitin-da** [comment]
Rebased)
